### PR TITLE
core(i18n): improve Intl polyfill and use it in Util

### PR DESCRIPTION
**Pull request by brendankenny**
fixes part of #7238 (second bullet point)

**tl;dr** better Intl polyfill guard and switch our Util number formatters to use `Intl` methods. No change to ICU message formatters.

Our current polyfill relies only on browserify build ignoring `intl`, which means we're polyfilling even on node that has `full-icu`. But if we base it just on if `Intl` and (e.g.) `Intl.NumberFormat` are defined we won't get locales on `system-icu` or `small-icu` where those are defined but only (usually) a single locale is supported.

Instead, we can use the approach suggested by [`intl-locales-supported`](https://www.npmjs.com/package/intl-locales-supported) and check if all our needed locales are supported and polyfilling if not. If we take that literally, though, we'll still polyfill because for some reason (TODO:) `'no'` and `'tl'` aren't included in the latest `Intl` release in Chrome/Node.

Instead we just check that a representative sample of locales are supported and, if so, count that Intl as up-to-date enough. I adopted the same set of locales that [`intl-pluralrules` uses](https://github.com/eemeli/intl-pluralrules/blob/30b37fce4621a1ce1865d1d4a0e2d1cf5ca98300/src/polyfill.js) but we could tweak to our own needs.

Finally, our `Util.format*()` methods are switched to use `new Intl.NumberFormat().format()` directly instead of indirectly through `Number.toLocaleString()`. This means that the Intl polyfill will be picked up in Node without `full-icu` and anywhere we were using those formatters (mostly in the renderer) will now be formatted. And we can re-enable some long-`skip`ped tests.

### CODE REVIEW

File: lighthouse-core/lib/i18n/i18n.js [hunk 1]
```diff
@@ -21,23 +21,25 @@ const MESSAGE_INSTANCE_ID_REGEX = /(.* \| .*) # (\d+)$/;
...
-  try {
-    // @ts-ignore
-    const IntlPolyfill = require('intl');
-    // In browser environments where we don't need the polyfill, this won't exist
-    if (!IntlPolyfill.NumberFormat) return;
+  // @ts-ignore
+  const IntlPolyfill = require('intl');
 
+  // Check if global implementation supports a minimum set of locales.
+  const minimumLocales = ['en', 'es', 'ru', 'zh'];
```
Comment (brendankenny) [inline lighthouse-core/lib/i18n/i18n.js]: this should be enough to establish whether Node is `full-icu` or not, but happy to add or take away from this minimum list.

File: lighthouse-core/report/html/renderer/util.js [hunk 2]
```diff
@@ -556,10 +555,11 @@ class Util {
...
    */
   static setNumberDateLocale(locale) {
-    Util.numberDateLocale = locale;
-
     // When testing, use a locale with more exciting numeric formatting
-    if (Util.numberDateLocale === 'en-XA') Util.numberDateLocale = 'de';
+    if (locale === 'en-XA') locale = 'de';
+
+    Util.numberDateLocale = locale;
+    Util.numberFormatter = new Intl.NumberFormat(locale);
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/util.js]: more static-ish state is unfortunate, but it seemed wasteful to do `new Intl.NumberFormat(Util.numberDateLocale).format(coarseTime)` for each format call...I'm not sure if it matters much

(tsc ensures it's initialized on the class, though, so that's a plus)

File: lighthouse-core/test/lib/i18n/swap-locale-test.js [hunk 3]
```diff
@@ -26,6 +26,12 @@ describe('swap-locale', () => {
     // Renderer formatted strings
     expect(lhrEn.i18n.rendererFormattedStrings.labDataTitle).toEqual('Lab Data');
     expect(lhrEs.i18n.rendererFormattedStrings.labDataTitle).toEqual('Datos de prueba');
+
+    // Formatted numbers in placeholders.
+    expect(lhrEn.audits['render-blocking-resources'].displayValue)
+      .toEqual('Potential savings of 1,130 ms');
+    expect(lhrEs.audits['render-blocking-resources'].displayValue)
+      .toEqual('Potential savings of 1.130 ms');
```
Comment (brendankenny) [inline lighthouse-core/test/lib/i18n/swap-locale-test.js]: this is actually formatted by `intl-messageformat`, but it's one more check that the number locales are being initialized correctly and it seemed like a missing piece of checking locale swapping

File: lighthouse-core/test/report/html/renderer/util-test.js [hunk 4]
```diff
@@ -47,25 +51,40 @@ describe('util helpers', () => {
     assert.equal(Util.formatDuration(28 * 60 * 60 * 1000 + 5000), `1${NBSP}d 4${NBSP}h 5${NBSP}s`);
   });
 
-  // TODO: need ICU support in node on Travis/Appveyor
-  it.skip('formats based on locale', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/report/html/renderer/util-test.js]: `-   it.skip`

🎉

File: lighthouse-core/lib/i18n/i18n.js [hunk 5]
```diff
@@ -21,23 +21,25 @@ const MESSAGE_INSTANCE_ID_REGEX = /(.* \| .*) # (\d+)$/;
 const MESSAGE_INSTANCE_ID_QUICK_REGEX = / # \d+$/;
 
 (() => {
-  // Node usually doesn't come with the locales we want built-in, so load the polyfill if we can.
+  // Node without full-icu doesn't come with the locales we want built-in. Load the polyfill if needed.
+  // See https://nodejs.org/api/intl.html#intl_options_for_building_node_js
+  // The bundler should remove these dependencies, so this will be a no-op in the browser.
```
Comment (patrickhulce) [inline lighthouse-core/lib/i18n/i18n.js]: might be worth clarifying "no-op". 

right now it appears to me that we would override a browser that failed to support those 4 locales with `Intl.NumberFormat = undefined`, which would be worse than a no-op. Is it more like `All our environments where IntlPolyfill is empty are guaranteed to support these 4 locales, so this will be a no-op`?

File: lighthouse-core/lib/i18n/i18n.js [hunk 6]
```diff
@@ -21,23 +21,25 @@ const MESSAGE_INSTANCE_ID_REGEX = /(.* \| .*) # (\d+)$/;
...
-    // In browser environments where we don't need the polyfill, this won't exist
-    if (!IntlPolyfill.NumberFormat) return;
+  // @ts-ignore
+  const IntlPolyfill = require('intl');
 
+  // Check if global implementation supports a minimum set of locales.
+  const minimumLocales = ['en', 'es', 'ru', 'zh'];
+  const supportedLocales = Intl.NumberFormat.supportedLocalesOf(minimumLocales);
+
+  if (supportedLocales.length !== minimumLocales.length) {
```
Comment (patrickhulce) [inline lighthouse-core/lib/i18n/i18n.js]: I'd still like to see some guard here if we're in the browser case
```suggestion
  if (supportedLocales.length !== minimumLocales.length && IntlPolyfill.NumberFormat) {
```

File: lighthouse-core/test/report/html/renderer/util-test.js [hunk 7]
```diff
@@ -9,6 +9,10 @@ const assert = require('assert');
 const Util = require('../../../../report/html/renderer/util.js');
 const sampleResult = require('../../../results/sample_v2.json');
 
+// Require i18n to make sure Intl is polyfilled in Node without full-icu.
```
Comment (patrickhulce) [inline lighthouse-core/test/report/html/renderer/util-test.js]: maybe worth noting that this code won't actually be run in node though for a real report but it's OK because browser should support it?

File: lighthouse-core/lib/i18n/i18n.js (hunk 5 above)
Comment (brendankenny) [reply lighthouse-core/lib/i18n/i18n.js]: Yeah, it shouldn't be possible in Chrome (other browsers never see this code because it's only in the bundle, not in the report) but you never know what Chrome builds are out there and at the least it would be good for the code to make it completely clear it's not possible. I'll fix it.

File: lighthouse-core/report/html/renderer/util.js [hunk 8]
```diff
@@ -213,7 +212,7 @@ class Util {
    */
   static formatMilliseconds(ms, granularity = 10) {
     const coarseTime = Math.round(ms / granularity) * granularity;
-    return `${coarseTime.toLocaleString(Util.numberDateLocale)}${NBSP}ms`;
+    return `${Util.numberFormatter.format(coarseTime)}${NBSP}ms`;
```
Comment (exterkamp) [inline lighthouse-core/report/html/renderer/util.js]: Out of scope, but disappointing that we just append 'ms' or 's' here.  We need that unit's proposal to go through!

Part of the process of making this consistent:
image

**Comment by brendankenny** [comment]
updated.

`Intl` now will never be touched if the modules have been emptied by browserify, followed by a second check that a minimum set of locales is supported before polyfilling in node.
